geopandas/sindex.py

```diff
@@ -70,6 +89,129 @@ def __init__(self, geometry):
+            self._geometries = geometry.geometry.values
+            # create a prepared geometry cache
+            self._prepared_geometries = np.array([None] * self._geometries.size)
+
+        def query(self, geometry, predicate=None, sort=False):
+            """Compatibility layer for pygeos.query.
+            This is not an optimized function, if speed is important,
+            you are probably better off using PyGEOS.
+            Parameters
+            ----------
+            geometry : shapely geometry
+                A single shapely geometry to query against the spatial index.
+            predicate : {None, 'intersects', 'within', 'contains', 'overlaps', 'crosses', 'touches'}, optional
+                If predicate is provided, a prepared version of the input geometry is tested using
+                the predicate function against each item in the index whose extent intersects the
+                envelope of the input geometry: predicate(geometry, tree_geometry).
+            sort : bool, default False
+                If True, the results will be sorted in ascending order. If False, results are
+                often sorted but there is no guarantee.
+            Returns
+            -------
+            matches : ndarray of shape (n_results, )
+                Integer indices for matching geometries from the spatial index.
+            """  # noqa: E501
+            # handle invalid predicates
+            if predicate not in self.valid_query_predicates:
+                raise ValueError(
+                    "Got `predicate` = `%s`; " % predicate
+                    + "`predicate` must be one of %s" % self.valid_query_predicates
+                )
+            # handle empty / invalid geometries
+            if geometry is None:
+                # this is the behavior in pygeos.strtree.query, we mimic it here
+                return np.array([], dtype=np.intp)
+            if not isinstance(geometry, BaseGeometry):
+                raise TypeError(
+                    "Got `geometry` of type `%s`, " % type(geometry)
+                    + "`geometry` must be a shapely geometry."
+                )
+            if geometry.is_empty:
+                return np.array([], dtype=np.intp)
+
+            # get bounds
+            bounds = geometry.bounds  # rtree operates on bounds
+            tree_query = list(self.intersection(bounds, objects=False))
+
+            if not tree_query:
+                return np.array([], dtype=np.intp)
+
+            # check predicate
+            if predicate in ("intersects", "within"):
+                # only contains and intersects are supported by
+                # prepared geometries, see note below regarding within
+                res = []
+                if predicate == "within":
+                    # since these are inverse, we can flip the operation
+                    # and test with prepared predicates from tree
+                    predicate = "contains"
```

Comment:
Shouldn't the geometries get flipped in order as well? (because it seems you are now using the same order for both contains and within ?)

Reply:
The logic here is:
```python3
if "within":
    tree[i].contains(geometry)
if "contains":
   geometry.contains(tree[i])
```

So indeed, the geometries do get flipped. What gets appended to the results is `i`, which in both cases is the index of the geometry in the tree. `query` only outputs indexes for tree geometries.

Reply:
I will try to add more comments to document this, if there is any specific way I can clarify this please let me know.

Reply:
I re-organized this section and added some more comments. Please let me know if it is clear now.

Reply:
Yeah, I missed they were using a different order in the actual call (I actually assumed this block was for within and contains, but it was actually for within and intersects, and contains was done below differently).

But added some new comments now ;)